File: B_2_D_Traveling.java
/*   कर्मण्येवाधिकारस्ते मा फलेषु कदाचन। मा कर्मफलहेतुर्भूर्मा ते सङ्गोऽस्त्वकर्माणि॥ 2-47॥   */

import java.util.Scanner;

public class Ques {
    private static class Two {
        long num1;
        long num2;
        Two(long num1, long num2) {
            this.num1 = num1;
            this.num2 = num2;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while(t --> 0) {
            int n = sc.nextInt();
            int k = sc.nextInt();
            int a = sc.nextInt();
            int b = sc.nextInt();

            Two[] arr = new Two[n];
            int i = -1;
            while(++i < n) arr[i] = new Two(sc.nextLong(), sc.nextLong());

            long startMinCost = Long.MAX_VALUE;
            long endMinCost = Long.MAX_VALUE;
            Two startFree;
            Two endFree;
            long ans = Math.abs(arr[a-1].num1-arr[b-1].num1) + Math.abs(arr[a-1].num2-arr[b-1].num2);
            if(k >= 2) {
                startFree = arr[0];
                endFree = arr[k-1];
                i = -1;
                while(++i < k) {
                    Two temp = arr[i];
                    long startCost = Math.abs(arr[a-1].num1-temp.num1) + Math.abs(arr[a-1].num2-temp.num2);
                    long endCost = Math.abs(arr[b-1].num1-temp.num1) + Math.abs(arr[b-1].num2-temp.num2);
                    if(startCost < startMinCost) {
                        startMinCost = startCost;
                        startFree = temp;
                    }
                    if(endCost < endMinCost) {
                        endMinCost = endCost;
                        endFree = temp;
                    }
                }
                ans = Math.min(ans, Math.abs(startFree.num1-arr[a-1].num1)+Math.abs(startFree.num2-arr[a-1].num2)+Math.abs(endFree.num1-arr[b-1].num1)+Math.abs(endFree.num2-arr[b-1].num2));
            }
            System.out.println(ans);
        }
    }
}

File: B_2_D_Traveling.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <cmath>
 
using namespace std;
 
class Two {
public:
    long long num1;
    long long num2;
    Two(long long num1, long long num2) : num1(num1), num2(num2) {}
};
 
int main() {
    int t;
    cin >> t;
    while (t-- > 0) {
        int n, k, a, b;
        cin >> n >> k >> a >> b;
 
        vector<Two> arr;
        for (int i = 0; i < n; ++i) {
            long long num1, num2;
            cin >> num1 >> num2;
            arr.emplace_back(num1, num2);
        }
 
        long long startMinCost = LLONG_MAX;
        long long endMinCost = LLONG_MAX;
        Two startFree(0, 0); // Initialize with default values
        Two endFree(0, 0);   // Initialize with default values
 
        // Cost of direct path from a to b
        long long ans = abs(arr[a-1].num1 - arr[b-1].num1) + abs(arr[a-1].num2 - arr[b-1].num2);
 
        if (k >= 2) {
            startFree = arr[0];
            endFree = arr[k-1];
            for (int i = 0; i < k; ++i) {
                Two temp = arr[i];
                long long startCost = abs(arr[a-1].num1 - temp.num1) + abs(arr[a-1].num2 - temp.num2);
                long long endCost = abs(arr[b-1].num1 - temp.num1) + abs(arr[b-1].num2 - temp.num2);
                if (startCost < startMinCost) {
                    startMinCost = startCost;
                    startFree = temp;
                }
                if (endCost < endMinCost) {
                    endMinCost = endCost;
                    endFree = temp;
                }
            }
            ans = min(ans, startMinCost + endMinCost);
        }
 
        cout << ans << endl;
    }
    return 0;
}